tfhe/src/typed_api/integers/server_key.rs

```diff
@@ -34,6 +34,169 @@ where
+    let lut = wopbs_key.generate_lut_bivariate_crt(&switched_lhs, &switched_rhs, func);
+    let res = wopbs_key.bivariate_wopbs_native_crt(&switched_lhs, &switched_rhs, lut.as_slice());
+    wopbs_key.keyswitch_to_pbs_params(&res)
+}
+
+pub trait WopbsEvaluationKey<ServerKey, Ciphertext> {
+    fn apply_wopbs(&self, sks: &ServerKey, ct: &Ciphertext, f: impl Fn(u64) -> u64) -> Ciphertext;
+
+    fn apply_bivariate_wopbs(
+        &self,
+        sks: &ServerKey,
+        lhs: &Ciphertext,
+        rhs: &Ciphertext,
+        f: impl Fn(u64, u64) -> u64,
+    ) -> Ciphertext;
+}
+
+impl
+    WopbsEvaluationKey<
+        crate::integer::ServerKey,
+        crate::typed_api::integers::server_key::RadixCiphertextDyn,
+    > for WopbsKey
+{
+    fn apply_wopbs(
+        &self,
+        sks: &crate::integer::ServerKey,
+        ct: &crate::typed_api::integers::server_key::RadixCiphertextDyn,
+        f: impl Fn(u64) -> u64,
+    ) -> crate::typed_api::integers::server_key::RadixCiphertextDyn {
+        match ct {
+            RadixCiphertextDyn::Big(ct) => {
+                let res = wopbs_radix(self, sks, ct, f);
+                RadixCiphertextDyn::Big(res)
+            }
+            RadixCiphertextDyn::Small(ct) => {
+                let res = wopbs_radix(self, sks, ct, f);
+                RadixCiphertextDyn::Small(res)
+            }
+        }
+    }
+
+    fn apply_bivariate_wopbs(
+        &self,
+        sks: &crate::integer::ServerKey,
+        lhs: &crate::typed_api::integers::server_key::RadixCiphertextDyn,
+        rhs: &crate::typed_api::integers::server_key::RadixCiphertextDyn,
+        f: impl Fn(u64, u64) -> u64,
+    ) -> crate::typed_api::integers::server_key::RadixCiphertextDyn {
+        match (lhs, rhs) {
+            (RadixCiphertextDyn::Big(lhs), RadixCiphertextDyn::Big(rhs)) => {
+                let res = bivariate_wopbs_radix(self, sks, lhs, rhs, f);
+                RadixCiphertextDyn::Big(res)
+            }
+            (RadixCiphertextDyn::Small(lhs), RadixCiphertextDyn::Small(rhs)) => {
+                let res = bivariate_wopbs_radix(self, sks, lhs, rhs, f);
+                RadixCiphertextDyn::Small(res)
+            }
+            (_, _) => {
+                unreachable!("internal error: cannot mix big and small ciphertext")
+            }
+        }
+    }
+}
+
+impl WopbsEvaluationKey<crate::integer::ServerKey, crate::integer::RadixCiphertextBig>
+    for WopbsKey
+{
+    fn apply_wopbs(
+        &self,
+        sks: &crate::integer::ServerKey,
+        ct: &crate::integer::RadixCiphertextBig,
+        f: impl Fn(u64) -> u64,
+    ) -> crate::integer::RadixCiphertextBig {
+        wopbs_radix(self, sks, ct, f)
+    }
+
+    fn apply_bivariate_wopbs(
+        &self,
+        sks: &crate::integer::ServerKey,
+        lhs: &crate::integer::RadixCiphertextBig,
+        rhs: &crate::integer::RadixCiphertextBig,
+        f: impl Fn(u64, u64) -> u64,
```

Comment:
can we perhaps already manage the RadixCtDyn here ? or do we need it given we manage the dyn thing above ?